app/javascript/lib/metrics-web-vitals.js

```diff
@@ -0,0 +1,141 @@
+ * Create Web Vitals Bard reporter, that accepts `Metric` values and sends it
+ * to Bard using `navigator.sendBeacon`.
+ *
+ * The function is called only once per page load.
+ *
+ * This is adapted from:
+ * https://github.com/treosh/web-vitals-reporter/blob/72a807e1df5749896668665d1d28867902ae909d/src/index.js
+ *
+ * @return {(metric: Metric) => void}
+ */
+export function createWebVitalsBardReporter() {
+  let isSent = false
+  let isCalled = false
+  let result = {}
+
+  const sendValues = () => {
+    if (isSent) {return} // data is already sent
+    if (!isCalled) {return} // no data collected
+
+    result.duration = performance.now()
+
+    result = Object.assign(result, hardwareStats)
+
+    isSent = true
+
+    if (typeof navigator === 'undefined') {return}
+    return log('web-vitals', result, true)
+  }
+
+
+  const mapMetric = function(metric) {
+    const isWebVital = ['FCP', 'TTFB', 'LCP', 'CLS', 'FID'].indexOf(metric.name) !== -1
+    return { [metric.name]: isWebVital ? round(metric.value, metric.name === 'CLS' ? 4 : 0) : metric.value }
+  }
+
+  /** @param {Metric} metric */
+  const report = metric => {
+    if (!isCalled) {isCalled = true}
+    result = { ...result, ...mapMetric(metric, result) }
+  }
+
+  // should be the last call to capture latest CLS
+  setTimeout(() => {
+    // Safari does not fire "visibilitychange" on the tab close
+    // So we have 2 options: loose Safari data, or loose LCP/CLS that depends on "visibilitychange" logic.
+    // Current solution: if LCP/CLS supported, use `onHidden` otherwise, use `pagehide` to fire the callback in the end.
+    //
+    // More details: https://github.com/treosh/web-vitals-reporter/issues/3
+    const supportedEntryTypes =
+      (typeof PerformanceObserver !== 'undefined' && PerformanceObserver.supportedEntryTypes) || []
+    const isLatestVisibilityChangeSupported = supportedEntryTypes.indexOf('layout-shift') !== -1
+
+    if (isLatestVisibilityChangeSupported) {
+      const onVisibilityChange = () => {
+        if (document.visibilityState === 'hidden') {
+          sendValues()
+          removeEventListener('visibilitychange', onVisibilityChange, true)
+        }
+      }
+      addEventListener('visibilitychange', onVisibilityChange, true)
+    } else {
+      addEventListener('pagehide', sendValues, { capture: true, once: true })
+    }
+  })
+
+  return report
+}
+
+/**
+ * Logs web vitals, performance metrics for page view UX, to Bard / Mixpanel
+ *
+ * Includes:
+ *  - TTFB: time to first byte, a very early page load event
+ *  - FCP: first contentful paint, when users gets first visual feedback
+ *  - LCP: largest contentful paint, a middle measure perceived load time
+ *  - FID: first input delay, when user can first interact (e.g. click)
+ *  - CLS: cumulative layout shift, measures visual stability
+ *  - Hardware stats on connection speed, # CPUs, memory
+ *
+ * LCP, FID, and CLS are the core web vitals.
+ * TTFB and FCP are useful supplements.
+ */
```

Comment:
do we have any sense for how long this call takes?  Right now we're blocking page load on this call completing, and I'm not sure that's what we want to do.  for performance, and error-safety, it might be better to execute this in a callback and/or try/catch

Reply:
Reading this PR more, it looks like this doesn't actually perform the logging, it just sets it up?  so maybe 'initializeWebVitalsLogging' is a more descriptive name?

Reply:
Yup, the setup call is non-blocking.  The log request happens when the page is *un*loaded, and also doesn't block.

This theoretically means that `web-vitals` log requests could get dropped.  The new-ish `sendBeacon` JS API is guarantees such requests wouldn't drop -- it was specifically designed to account for analytics use cases like this.   However, Bard requires an `Authorization: Bearer` HTTP request header for registered users, and `sendBeacon` doesn't support such non-standard request headers.  

So until Bard is enhanced to authenticate user identities some other way, this approach is the next best thing.

Reply:
`b80cace` uses clearer function names.

Reply:
cool